### PR TITLE
Enable @typescript-eslint/no-shadow and @typescript-eslint/no-redeclare and updated the project

### LINKED ISSUE #4219: Enable @typescript-eslint/no-shadow and @typescript-eslint/no-redeclare for the project
## Feature Suggestion
Now after #4210 is merged, we should enabled the two following rules in our `cli/config/eslint*.js` files.

```js
{
  rules: {
    '@typescript-eslint/no-shadow': 'error',
    '@typescript-eslint/no-redeclare': 'error'
  }
}
```

And then update the project to work with these new rules :)

### PR DESCRIPTION
**Pull request by ron-debajyoti**
## Hey, I just made a Pull Request!

Closes #4219 by updating the new lint rules ( `no-shadow` and `no-redeclare` ) and then updating the required project files. Almost all the file changes are renaming the variable names.

#### :heavy_check_mark: Checklist

- [ ]  A changeset describing the change and affected packages
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Review by OrkoHunter — APPROVED** [review]
Looks good for TechDocs.

**Comment by OrkoHunter** [comment]
Should rebase with the latest master branch for the conflict.

**Comment by freben** [comment]
@ron-debajyoti Just a gentle ping :) Is it going well?

File: packages/core-api/src/lib/AuthConnector/DirectAuthConnector.ts [hunk 1]
```diff
@@ -27,11 +27,11 @@ type Options = {
...
-export type DirectAuthResponse = {
-  userId: string;
-  profile: ProfileInfo;
-  backstageIdentity: BackstageIdentity;
-};
+// export type DirectAuthResponse = {
+//   userId: string;
+//   profile: ProfileInfo;
+//   backstageIdentity: BackstageIdentity;
+// };
```
Comment (freben) [inline packages/core-api/src/lib/AuthConnector/DirectAuthConnector.ts:34]: This seems to be accidental

File: packages/core/src/layout/HeaderTabs/HeaderTabs.tsx [hunk 2]
```diff
@@ -38,13 +38,13 @@ const useStyles = makeStyles(theme => ({
   },
 }));
 
-export type Tab = {
+export type objectTab = {
```
Comment (freben) [inline packages/core/src/layout/HeaderTabs/HeaderTabs.tsx:41]: Types should always be uppercase. Also this type is actually exported so it's part of the contract - can we keep this  the way it was, and adapt elsewhere instead?

File: packages/integration/src/github/GithubCredentialsProvider.test.ts [hunk 3]
```diff
@@ -215,21 +215,21 @@ describe('GithubCredentialsProvider tests', () => {
   });
 
   it('should return the default token if no app is configured', async () => {
-    const github = GithubCredentialsProvider.create({
+    const githubID = GithubCredentialsProvider.create({
```
Comment (freben) [inline packages/integration/src/github/GithubCredentialsProvider.test.ts:218]: maybe call it `provider`?

File: plugins/cost-insights/src/components/BarChart/BarChartLabel.tsx [hunk 4]
```diff
@@ -18,7 +18,7 @@ import React, { PropsWithChildren } from 'react';
 import { Box, Typography } from '@material-ui/core';
 import { useBarChartLabelStyles } from '../../utils/styles';
 
-type BarChartLabel = {
+type BarChartLabelType = {
```
Comment (freben) [inline plugins/cost-insights/src/components/BarChart/BarChartLabel.tsx:21]: really? this seems like an odd rename

File: plugins/cost-insights/src/components/BarChart/BarChartSteps.tsx [hunk 5]
```diff
@@ -18,7 +18,7 @@ import React from 'react';
 import { ButtonBase } from '@material-ui/core';
 import { useBarChartStepperStyles as useStyles } from '../../utils/styles';
 
-export type BarChartSteps = {
+export type BarChartStepsType = {
```
Comment (freben) [inline plugins/cost-insights/src/components/BarChart/BarChartSteps.tsx:21]: same here

File: plugins/cost-insights/src/components/ProductInsights/ProductInsights.tsx [hunk 6]
```diff
@@ -83,28 +83,31 @@ export const ProductInsights = ({
 
   useEffect(() => {
     async function getAllProductInsights(
-      group: string,
-      project: Maybe<string>,
-      products: Product[],
-      lastCompleteBillingDate: string,
+      groupId: string,
+      projectId: Maybe<string>,
+      productsId: Product[],
```
Comment (freben) [inline plugins/cost-insights/src/components/ProductInsights/ProductInsights.tsx:88]: `productIds`?

File: plugins/cost-insights/src/components/ProductInsights/ProductInsights.tsx [hunk 7]
```diff
@@ -83,28 +83,31 @@ export const ProductInsights = ({
...
+      lastCompleteBillingDateId: string,
```
Comment (freben) [inline plugins/cost-insights/src/components/ProductInsights/ProductInsights.tsx:89]: mmh not sure about this one, maybe `lastCompleteBillingDateString`

File: plugins/scaffolder/src/components/MultistepJsonForm/MultistepJsonForm.tsx [hunk 8]
```diff
@@ -30,7 +30,7 @@ import { Theme as MuiTheme } from '@rjsf/material-ui';
 import React, { useState } from 'react';
 
 const Form = withTheme(MuiTheme);
-type Step = {
+type StepType = {
```
Comment (freben) [inline plugins/scaffolder/src/components/MultistepJsonForm/MultistepJsonForm.tsx:33]: really is super unfortunate to add Type to types, that kind of naming should be avoided - where does this colide?

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx [hunk 9]
```diff
@@ -122,8 +124,8 @@ export const TemplatePage = () => {
 
   const handleCreate = async () => {
     try {
-      const jobId = await scaffolderApi.scaffold(templateName, formState);
-      setJobId(jobId);
+      const Id = await scaffolderApi.scaffold(templateName, formState);
```
Comment (freben) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:127]: lowercase

File: plugins/search/src/components/SearchResult/SearchResult.tsx [hunk 10]
```diff
@@ -59,10 +59,10 @@ type TableHeaderProps = {
...
 };
 
-type Filters = {
-  selected: string;
-  checked: Array<string | null>;
-};
+// type Filters = {
+//   selected: string;
+//   checked: Array<string | null>;
+// };
```
Comment (freben) [inline plugins/search/src/components/SearchResult/SearchResult.tsx:65]: mistake?

**Comment by ron-debajyoti** [comment]
Hey @freben !, I fixed most of the typos, still have a few issues.

**Comment by ron-debajyoti** [comment]
Also, do I have to add a changeset?

**Comment by timbonicus** [comment]
I had some minor changes to the `cost-insights` files, but didn't see a nice way to push to the fork. Can you apply this patch?


[pr-4300-cost-insights.patch.zip](https://github.com/backstage/backstage/files/5912642/pr-4300-cost-insights.patch.zip)

**Comment by ron-debajyoti** [comment]
@timbonicus Are the changes related to #4219

**Comment by timbonicus** [comment]
@ron-debajyoti Yes, the patch is directly related to changes in this PR. You can apply it after unzipping with `git apply pr-4300-cost-insights.patch`

**Comment by ron-debajyoti** [comment]
@timbonicus thanks a lot! Committing them asap.

**Review by timbonicus — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]
Thanks! 👍

**Comment by OrkoHunter** [comment]
> I had some minor changes to the `cost-insights` files, but didn't see a nice way to push to the fork. Can you apply this patch?

@timbonicus I use the [GitHub CLI](https://cli.github.com/manual/gh_pr_checkout) to checkout _any_ PR. :)

```
gh pr checkout 4300
# commit
git push
```

**Comment by ron-debajyoti** [comment]
Hi !
Can we get a final review and merge this PR quickly to avoid frequent merge conflicts ?

**Comment by benjdlambert** [comment]
Think we missed a changeset for this. We should include one and suggest how people can turn it off similar to last release's eslint change too. https://github.com/backstage/backstage/commit/12a56cdfe0ac6875cc992167079e8215ce476df7

**Comment by freben** [comment]
thanks for noting that, i'll make one separately

**Pull request merged**